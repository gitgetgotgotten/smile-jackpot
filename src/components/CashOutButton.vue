<template>
  <div id="cash-out-wrapper" @mouseenter="moveButton" @mouseleave="resetButtonPosition">
    <button ref="cashOutEl" id="cash-out" @click="cashOut">CASH OUT</button>
  </div>
</template>

<script>
import {ref} from 'vue'

export default {
  emits: ['cashedOut'],

  setup(_, {emit}) {
    const cashOutEl = ref(null)

    /**
     * Prevent the `cashOut` from working properly by making the button un-clickable in order to keep the cash
     */
    const cashOut = () => {
      if (Math.random() > 0.4) {
        emit('cashedOut');
      }
    }
    /**
     * Prevent the `cashOut` from working properly by moving the button to a random position in order to keep the cash
     */
    const moveButton = () => {
      if (Math.random() <= 0.5) {
        const randomDistance = 300;
        const randomDirection = Math.floor(Math.random() * 4) + 1;
        if (randomDirection === 1) {
          cashOutEl.value.style.transform = `translateY(-${randomDistance}px)`;
        } else if (randomDirection === 2) {
          cashOutEl.value.style.transform = `translateY(${randomDistance}px)`;
        } else if (randomDirection === 3) {
          cashOutEl.value.style.transform = `translateX(-${randomDistance}px)`;
        } else if (randomDirection === 4) {
          cashOutEl.value.style.transform = `translateX(${randomDistance}px)`;
        }
      }
    }
    /**
     * Reset the position of the button to its original place
     */
    const resetButtonPosition = () => {
      cashOutEl.value.style.transform = '';
    }

    return {
      cashOutEl,
      cashOut,
      moveButton,
      resetButtonPosition,
    }
  },
};
</script>

<style scoped>
#cash-out-wrapper {
  display: inline-block;
}
</style>