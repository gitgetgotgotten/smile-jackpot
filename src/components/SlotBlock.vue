<template>
  <div class="slot-block">
    <img v-if="spinning" alt="arrow down" src="/src/assets/icons/spinner.svg" width="50" height="50"/>
    <img v-else :id="icon" :alt="icon" :src="`/src/assets/icons/${icon}.svg`" width="50" height="50"/>
  </div>
</template>

<script>
export default {
  props: {
    spinning: {
      type: Boolean,
      required: true
    },
    icon: {
      type: String,
      required: true
    },
  },
};
</script>

<style scoped>
.slot-block {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100px;
  height: 100px;
  border: 1px solid #000;
}
</style>