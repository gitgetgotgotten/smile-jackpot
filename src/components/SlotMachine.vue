<template>
  <div>
    <div class="slot-machine">
      <SlotBlock v-for="(_, i) in 3" :key="i" :icon="rolledBlocks[i].icon" :spinning="spinning[i]"/>
      <button id="roll-button" :disabled="isSpinning || credits === 0" @click="roll">
        {{ isSpinning ? 'Rolling...' : 'Roll' }}
      </button>
    </div>
    <CashOutButton @cashedOut="settlement"/>
  </div>
</template>

<script>
import SlotBlock from './SlotBlock.vue';
import CashOutButton from './CashOutButton.vue';
import {ref, reactive} from 'vue';

export default {
  components: {
    SlotBlock,
    CashOutButton,
  },
  setup() {
    const tinyReRollChance = 0.3;
    const largeReRollChance = 0.6;
    const tinyCredits = 40;
    const mediumCredits = 60;
    const credits = ref(10);
    const user = reactive({
      credits: 0
    });

    const spinning = reactive([false, false, false]);
    const isSpinning = ref(false);

    const blocks = [
      {symbol: 'C', reward: 10, icon: 'cherry'},
      {symbol: 'L', reward: 20, icon: 'lemon'},
      {symbol: 'O', reward: 30, icon: 'orange'},
      {symbol: 'W', reward: 40, icon: 'watermelon'}
    ];

    const rolledBlocks = reactive([
      {icon: 'arrow-down'},
      {icon: 'arrow-down'},
      {icon: 'arrow-down'},
    ]);

    /**
     * Transfer what the users have in their `credits` to their user account credits
     */
    const settlement = () => {
      const creditsValue = credits.value;
      credits.value = 0;
      user.credits += creditsValue;
    };

    /**
     * Start the game, decrease from the user's credits, and roll the blocks
     */
    const roll = () => {
      isSpinning.value = true;
      spinning.forEach((_, index) => (spinning[index] = true));
      rollBlocks();
      calculateResult();
    };

    /**
     * Get a random symbol for each block
     */
    const rollBlocks = () => {
      rolledBlocks.forEach((_, index) => {
        const randomIndex = Math.floor(Math.random() * blocks.length);
        rolledBlocks[index] = blocks[randomIndex];
      });
    };

    /**
     * Check if all blocks match, needs to roll again?, and increase the credits if the blocks match
     */
    const calculateResult = () => {
      const allBlocksMatch =
        rolledBlocks[0].symbol === rolledBlocks[1].symbol &&
        rolledBlocks[1].symbol === rolledBlocks[2].symbol;

      if (allBlocksMatch) {
        if (credits.value < tinyCredits || rollAgainChance() === false) {
          credits.value += rolledBlocks[0].reward;
          showResult();
        } else {
          roll();
        }
      } else {
        credits.value--;
        showResult();
      }
    };

    /**
     * Check if it needs to roll again by chance
     * @returns {boolean}
     */
    const rollAgainChance = () => {
      let rollAgain = false;
      if (credits.value >= tinyCredits && credits.value <= mediumCredits) {
        rollAgain = Math.random() <= tinyReRollChance;
      } else if (credits.value > mediumCredits) {
        rollAgain = Math.random() <= largeReRollChance;
      }

      return rollAgain;
    };

    /**
     * Show the last result in the blocks
     */
    const showResult = () => {
      // Show 1st block after 1 second
      setTimeout(() => {
        spinning[0] = false;
      }, 1000);

      // Show 2nd block after 2 seconds
      setTimeout(() => {
        spinning[1] = false;
      }, 2000);

      // Show 3rd block after 3 seconds
      setTimeout(() => {
        spinning[2] = false;
        isSpinning.value = false;
      }, 3000);
    };

    return {
      tinyReRollChance,
      largeReRollChance,
      tinyCredits,
      mediumCredits,
      credits,
      user,
      spinning,
      isSpinning,
      blocks,
      rolledBlocks,
      settlement,
      roll,
      showResult
    };
  }
};

</script>

<style scoped>
.slot-machine {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom: 1rem;
}
</style>