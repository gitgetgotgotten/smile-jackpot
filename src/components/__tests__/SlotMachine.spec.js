import {describe, test, expect} from 'vitest';
import {mount} from '@vue/test-utils';
import SlotMachine from '/src/components/SlotMachine.vue';
import SlotBlock from '/src/components/SlotBlock.vue';
import CashOutButton from '/src/components/CashOutButton.vue';

describe('SlotMachine', () => {
  // To be used inside some test()s
  const wrapper = mount(SlotMachine);

  test('should load <SlotMachine>', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('should have 3 <SlotBlock>s', () => {
    const slotBlocks = wrapper.findAllComponents(SlotBlock);
    expect(slotBlocks.length).toBe(3);
  });

  test('should have 1 <CashOutButton>', () => {
    const cashOutButton = wrapper.findAllComponents(CashOutButton);
    expect(cashOutButton.length).toBe(1);
  });

  test('should have a roll button', () => {
    const rollButton = wrapper.find('button#roll-button');
    expect(rollButton.exists()).toBe(true);
    expect(rollButton.element.id).toBe('roll-button');
  });

  test('should have correct data', () => {
    expect(wrapper.vm.tinyReRollChance).toBe(0.3);
    expect(wrapper.vm.largeReRollChance).toBe(0.6);
    expect(wrapper.vm.tinyCredits).toBe(40);
    expect(wrapper.vm.mediumCredits).toBe(60);
    expect(wrapper.vm.credits).toEqual(10);
    expect(wrapper.vm.user).toEqual({credits: 0});
    expect(wrapper.vm.spinning).toEqual([false, false, false]);
    expect(wrapper.vm.isSpinning).toBe(false);
    expect(wrapper.vm.blocks).toEqual([
      {symbol: 'C', reward: 10, icon: 'cherry'},
      {symbol: 'L', reward: 20, icon: 'lemon'},
      {symbol: 'O', reward: 30, icon: 'orange'},
      {symbol: 'W', reward: 40, icon: 'watermelon'},
    ]);
    expect(wrapper.vm.rolledBlocks).toEqual([
      {icon: 'arrow-down'},
      {icon: 'arrow-down'},
      {icon: 'arrow-down'},
    ]);
  });

  test('should transfer `credits` to user\'s credits', () => {
    const wrapper = mount(SlotMachine);
    wrapper.vm.user.credits = 100;
    wrapper.vm.settlement();
    expect(wrapper.vm.user.credits).toBe(110);
  });

  test('should change spinning states and have valid rolledBlocks after running roll()', () => {
    const wrapper = mount(SlotMachine);
    const tempRolledBlocks = wrapper.vm.rolledBlocks;

    // Spinning changed to true
    wrapper.vm.roll();
    expect(wrapper.vm.isSpinning).toBe(true);
    expect(wrapper.vm.spinning).toEqual([true, true, true]);

    // Blocks exist in blocks and are valid
    const exists = wrapper.vm.rolledBlocks.some((obj, index) => {
      const tempRolledBlock = tempRolledBlocks[index];
      return obj.symbol === tempRolledBlock.symbol &&
        obj.reward === tempRolledBlock.reward &&
        obj.icon === tempRolledBlock.icon;
    });
    expect(exists).toBe(true);
  });

  test('should increase credits depending on the symbol reward when the symbols match and user has less than 40 credits', () => {
    const wrapper = mount(SlotMachine);
    const tempCredits = wrapper.vm.credits;
    wrapper.vm.roll();

    const rolledBlocks = wrapper.vm.rolledBlocks;
    if (
      rolledBlocks[0].symbol === rolledBlocks[1].symbol &&
      rolledBlocks[1].symbol === rolledBlocks[2].symbol
    ) {
      if (wrapper.vm.credits < wrapper.vm.credits.tinyCredits) {
        expect(wrapper.vm.credits).toBe(tempCredits + rolledBlocks[0].reward);
      }
    }
  });

  test('should decrement 1 unit from credits when the symbols don\'t match', () => {
    const wrapper = mount(SlotMachine);
    const tempCredits = wrapper.vm.credits;
    wrapper.vm.roll();

    const rolledBlocks = wrapper.vm.rolledBlocks;
    if (
      rolledBlocks[0].symbol !== rolledBlocks[1].symbol ||
      rolledBlocks[1].symbol !== rolledBlocks[2].symbol
    ) {
      expect(wrapper.vm.credits).toBe(tempCredits - 1);
    }
  });
});