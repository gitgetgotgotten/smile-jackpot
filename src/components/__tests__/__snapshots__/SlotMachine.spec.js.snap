// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SlotMachine > renders the component correctly 1`] = `
"<div data-v-e16b67b2=\\"\\">
  <div data-v-e16b67b2=\\"\\" class=\\"slot-machine\\">
    <slot-block-stub data-v-e16b67b2=\\"\\" icon=\\"arrow-down\\" spinning=\\"false\\"></slot-block-stub>
    <slot-block-stub data-v-e16b67b2=\\"\\" icon=\\"arrow-down\\" spinning=\\"false\\"></slot-block-stub>
    <slot-block-stub data-v-e16b67b2=\\"\\" icon=\\"arrow-down\\" spinning=\\"false\\"></slot-block-stub><button data-v-e16b67b2=\\"\\" class=\\"roll-button\\">Roll</button>
  </div>
  <cash-out-button-stub data-v-e16b67b2=\\"\\"></cash-out-button-stub>
</div>"
`;
